core/example/benchmark/Cabana_CommPerformance.cpp

```diff
@@ -0,0 +1,573 @@
+    std::vector<int> ranks_per_dim( 3 );
+    MPI_Dims_create( comm_size, 3, ranks_per_dim.data() );
+
+    // Generate a communicator with a cartesian topology and periodic
+    // boundaries.
+    std::vector<int> periodic_dims( 3, 1 );
+    MPI_Comm cart_comm;
+    int reorder_cart_ranks = 1;
+    MPI_Cart_create( comm,
+                     3,
+                     ranks_per_dim.data(),
+                     periodic_dims.data(),
+                     reorder_cart_ranks,
+                     &cart_comm );
+
+    // Get the Cartesian topology index of this rank.
+    int linear_rank;
+    MPI_Comm_rank( cart_comm, &linear_rank );
+    std::vector<int> cart_rank( 3 );
+    MPI_Cart_coords( cart_comm, linear_rank, 3, cart_rank.data() );
+
+    // Compute the 27 ranks this rank communicates with. This includes self
+    // communication. Put self communication first. This list needs to be
+    // unique.
+    std::vector<int> neighbor_ranks;
+    neighbor_ranks.push_back( linear_rank );
+    for ( int k = -1; k < 2; ++k )
+        for ( int j = -1; j < 2; ++j )
+            for ( int i = -1; i < 2; ++i )
+                if ( !(i==0 && j==0 && k==0) )
+                {
+                    std::vector<int> ncr = { cart_rank[0] + i,
+                                             cart_rank[1] + j,
+                                             cart_rank[2] + k };
+                    int nr;
+                    MPI_Cart_rank( cart_comm, ncr.data(), &nr );
+                    neighbor_ranks.push_back( nr );
+                }
+    std::sort( neighbor_ranks.begin(), neighbor_ranks.end() );
+    auto unique_end =
+        std::unique( neighbor_ranks.begin(), neighbor_ranks.end() );
+    neighbor_ranks.resize( std::distance(neighbor_ranks.begin(),unique_end) );
+
+    // Define the aosoa.
+    using comm_memory_space = typename CommDevice::memory_space;
+    using data_memory_space = typename DataDevice::memory_space;
+    using member_types = Cabana::MemberTypes<double[3],
+                                             double[3],
+                                             double,
+                                             int>;
+    using aosoa_type = Cabana::AoSoA<member_types,data_memory_space>;
+
+    // Get the byte size of a particle
+    using tuple_type = Cabana::Tuple<member_types>;
+    int bytes_per_particle = sizeof( tuple_type );
+
+    // Number of runs in the test loops.
+    int num_run = 10;
+
+    // Fraction of particles on each rank that will be communicated to the
+    // neighbors. We will sweep through these fractions to get an indicator of
+    // performance as a function of message size.
+    std::vector<double> comm_fraction = { 0.0001,
+                                          0.001,
+                                          0.005,
+                                          0.01,
+                                          0.05,
+                                          0.10,
```

Comment:
Should we add something between `0.1` and `0.5`? For PIC it's not unbelievable to get something a little higher than `0.1` (although between ~3-10% if typical)

Reply:
we could add 0.25 here. You really want to have a realistic number of particles per GPU for this test. You need to test both message size as well as performance of the other kernels.

Reply:
done